pkg/rpcserver/rpcserver.go

```diff
@@ -278,22 +295,44 @@ func (serv *server) handleRunnerConn(runner *Runner, conn *flatrpc.Conn) error {
 		opts.Features = serv.cfg.Features
 	}
 
-	err := runner.Handshake(conn, opts)
+	info, err := runner.Handshake(conn, opts)
 	if err != nil {
 		log.Logf(1, "%v", err)
 		return err
 	}
 
+	eg, ctx := errgroup.WithContext(baseCtx)
+
+	serv.checkOnce.Do(func() {
+		// Run the machine check.
+		eg.Go(func() error {
+			if err := serv.runCheck(ctx, &info); err != nil {
+				return fmt.Errorf("%w: %w", errFatal, err)
+			}
+			return nil
+		})
+	})
+
 	if serv.triagedCorpus.Load() {
-		if err := runner.SendCorpusTriaged(); err != nil {
-			log.Logf(2, "%v", err)
-			return err
-		}
+		eg.Go(runner.SendCorpusTriaged)
 	}
 
-	return serv.connectionLoop(runner)
+	stopCtx, stop := context.WithCancel(ctx)
+	go func() {
+		<-stopCtx.Done()
+		runner.Stop()
+	}()
+	defer stop()
```

Comment:
This looks fishy. stop will run after eg.Wait(), so if it's supposed to cancel the above goroutine, it won't.

Reply:
It will work as intended in the case then `ctx` itself is cancelled -- it will call `runner.Stop()` and exit. The `defer stop()` part doesn't seem to be necessary indeed -- if we the loop stopped, we don't really care if `runner.Stop()` is called after that anyway.

Rewrote to only rely on `ctx`.